### PR TITLE
New: convergence and divergence rules

### PR DESCRIPTION
**Pull request by nimo23**
Changes proposed in this pull request:
- ConvergenceDivergenceIndicator.java
- ChaikinOscillator.java
- add parameter "name" in (Base)Strategy.java

- [x] added an entry to the unreleased section of `CHANGES.md`

### CODE REVIEW

**Comment by nimo23** [comment]
Please review code and check for **logical failures especially for the "otherIsNotHighest"** in "IsDivergent/IsConvergentPositiveRule" and **"refIsNotHighest"** in "IsDivergent/IsConvergentNegativeRule". 

I guess, with "otherIsNotHighest" and "refIsNotHighest", the "delay" between the two compared indicators are recognized: For example, even one makes higher highs (e.g. oscillator) and the other (e.g. price) does not follow this with the same speed, IsDivergentRule will recognize it as divergence (which is fine). Maybe you find improvements.

**Comment by nimo23** [comment]
Maybe I can use the "CorrelationCoefficientIndicator.java" from ta4j/statistics within the convergent/divergentRules:

- if Correlation Coefficient>0, the two indicators move together (->convergent). 
- if Correlation Coefficient<0, the two indicators move in opposite directions (->convergent). 

 What do you think?

**Comment by team172011** [comment]
I am unsure about the current `useStrictConvergence=true` case. Using statistics like `Correlation Coefficient` for that case sounds good

**Comment by nimo23** [comment]
If useStrictConvergence=true, then CorrelationCoefficientIndicator is not used, otherwise it is used.

Look above at the example.

**Comment by nimo23** [comment]
Which one should I use:

append with AND:
` .and(ccIsGt0);`

append with OR:
` .or(ccIsGt0);`

append with XOR:
` .xor(ccIsGt0);`

**Comment by team172011** [comment]
I am unsure about the logical link in gerneral, because i can imagine several cases wherer the rule would be wrong satisfied:
For example `IsConvergentNegativeRule `
- `useStrictConvergence = true` -> "if both are falling" `satisfied = true` -> i am ok with that
- `useStrictConvergence = false`

| ref | other | case | possible situation |
| ----------|---------- |-------|-----------|
| ----------|---------- |-------|-----------|

**Comment by nimo23** [comment]
Yes, you are right. I will use solely "CorrelationCoefficientIndicator". 

However, how should I determine the direction (**positive/negative**)? Should I use another Trend-Indicator/Oscillator to determine the direction of "ref" and the direction of "other". For example:

- 1.step: CorrelationCoefficientIndicator determines that ref and other is divergent
- 2.step: MovingAvgCrossover determines that "ref" is going up
- 3.step: MovingAvgCrossover determines that "other" is going down
--> this means: **positive** divergence


 Or does any other statistical function exist to determine the direction?

File: ta4j-core/src/main/java/org/ta4j/core/Strategy.java [hunk 1]
```diff
@@ -83,8 +88,7 @@ default boolean shouldEnter(int index, TradingRecord tradingRecord) {
         if (isUnstableAt(index)) {
             return false;
         }
-        final boolean enter = getEntryRule().isSatisfied(index, tradingRecord);
```
Comment (nimo23) [inline ta4j-core/src/main/java/org/ta4j/core/Strategy.java]: No need for "final boolean enter".

File: ta4j-core/src/main/java/org/ta4j/core/Strategy.java [hunk 2]
```diff
@@ -104,7 +108,6 @@ default boolean shouldExit(int index, TradingRecord tradingRecord) {
         if (isUnstableAt(index)) {
             return false;
         }
-        final boolean exit = getExitRule().isSatisfied(index, tradingRecord);
```
Comment (nimo23) [inline ta4j-core/src/main/java/org/ta4j/core/Strategy.java]: No need for "final boolean exit".

**Comment by nimo23** [comment]
I accidently changed the Strategy/BaseStrategy within this PR instead of making a new PR. I only added a optional parameter "name" to be able to give a name for the strategy.

**Comment by nimo23** [comment]
"IsConvergentNegative" will have the following logic:

```
@Override
	public boolean isSatisfied(int index, TradingRecord tradingRecord) {
		
		CorrelationCoefficientIndicator cc = new CorrelationCoefficientIndicator(ref, other, timeFrame);
		boolean isConvergent = cc.getValue(index).isGreaterThanOrEqual(Decimal.valueOf(minStrenght).abs());
		
		HMAIndicator hma = new HMAIndicator(ref, timeFrame);
		boolean isNegative = ref.getValue(index).isLessThan(hma.getValue(index));
		
		final boolean satisfied = isConvergent && isNegative;
		traceIsSatisfied(index, satisfied);
		return satisfied;
	}
```

Maybe we can use something better than HMAIndicator to find the direction?

**Comment by nimo23** [comment]
**The only problem with the logic is:**

Can we use something better than HMAIndicator to find the direction? Or is SimpleMovingAverage enough? Other TrendDirection-Indicators? Maybe with the help of ta4j-"SimpleLinearRegression.java" ?

**Comment by team172011** [comment]
To detect a strict convergence/divergence is easy, but to detect some kind of "not so strict" convergence/divergence could be very vague. One user would use an EMA indicator another user would like to use HAMA and the third one would use Arron.... 
In my opining it is to specific to to identify the trend with the help of a specific indicator in a rule. A rule should be **as general as possible**. So one of the commonest detection could be like you say through a **simple linear regression** over the given time frame. To specifiy how strict this "not so strict" case can be we could use a parameter to compare the slope of the regression for each indicator. 

For example a value between 0 (not strict) - 100 (very strict)

- `parameter == 100` -> slope mus be exact the same

- `parameter == 50` -> slope of ref can be half of slop of other or more...

**Comment by nimo23** [comment]
When I have this:

`SimpleLinearRegressionIndicator reg = new SimpleLinearRegressionIndicator(close, 7);`

and it outputs this for index n:

```
closeprice(n): "27.82"
simpleLinearReg(n): "27.7761"
```

**does this mean the following ?:**

For index n the n+1 value is probably around "27.7761" (= prediction value).

**Does "SimpleLinearRegressionIndicator" outputs the "predictionValue" for the next comming index?**

**Comment by nimo23** [comment]
I should only look at the slope of each regression value?

**For example:**

```
(1) closeprice=27.68, regression=null, slope=null
(2) closeprice=27.66, regression=27.6643, slope= 27.6643
(3) closeprice=27.66, regression=27.6589, slope=-0.0054
(4) closeprice=27.67, regression=27.6567, slope=-0.0022
(5) closeprice=27.64, regression=27.656, slope=-0.0007 = (regression(5)) - (regression(4))
```
**With a timeFrame=3 on index (5), I can say:**
(regression(5)) - (regression(2)) = -0.0083 **(downtrend?)**

Should I assume that "-0.0083" is saying: negative->downtrend ?

**Comment by team172011** [comment]
Spontaneous i would suggest to compare the slope of the regression line over the time frame:

grafik

**Comment by nimo23** [comment]
When I have this:
I updated "SimpleLinearRegressionIndicator" to be able to set the returned type of the value (slope, intercept, y) within an optional constructor.

Now I can do this:

```
// timeFrame=7
SimpleLinearRegressionIndicator reg = new SimpleLinearRegressionIndicator(close, 7, SimpleLinearRegressionType.slope);
```

and it returns for index(n) the slope=-0.0058.

Is the value (-0.0058) **the slope between "timeFrame-7" till "timeFrame"** or is this **the slope between index 7 and its previous index 6** ?

**Comment by team172011** [comment]
I have to take a look at your implementation of `SimpleLinearRegressionIndicator`, but you could check with

- slope between (8) and (1) = (27.6567-27.735)/(8-1) = -0,0112

- slope between (8) and (7) =  (27.6567-27.6589)/(8-7) = -022

**Comment by nimo23** [comment]
I did not change the calculation of "SimpleLinearRegressionIndicator"..

**Comment by nimo23** [comment]
I did not change the calculation of SimpleLinearRegressionIndicator. I only output the "slope" instead of "y" by setting the type within constructor. 

Why did you divide to "/(8-2)"?

The slope is nothing more than the difference of x and y and can also be calculated by DifferenceIndicator. Am I right?

So why using the slope of SimpleLinearRegression? Is there someting different?

**Comment by nimo23** [comment]
By the way, actually the InSlopeRule only cares about the difference of two points and treats this as a slope. Should I change InSlopeRule to calculate the slope with the help of SimpleLinearRegression or the formula above you provided?

**Comment by nimo23** [comment]
I need the "relative slope" instead of the "absolute slope" (which is calculated above) to be able to compare to "minSlope"-parameter.

**For example, a value between 0 (not strict) - 100 (very strict):**
- parameter >= +minSlope -> positive slope for positiveDivergence/positiveConvergence.
- parameter <= -minSlope -> negative slope for negativeDivergence/negativeConvergence.

**Comment by nimo23** [comment]
I have (almost?) finished the "ConvergenceDivergenceIndicator.java". 

**Please review the source.** 

After that I am actually writing tests for "ConvergenceDivergenceIndicatorTest.java" for all cases.

**Comment by team172011** [comment]
> By the way, actually the InSlopeRule.java only cares about the difference of two points and treats this as a slope. Should I change InSlopeRule.java to calculate the slope with the help of SimpleLinearRegression or the above formula you provided?

Good question, but for the moment i think the rule is regarding its javadoc description ok.

File: ta4j-core/src/main/java/org/ta4j/core/BaseStrategy.java [hunk 3]
```diff
@@ -53,25 +56,43 @@ this software and associated documentation files (the "Software"), to deal in
...
     
     /**
      * Constructor.
+     * @param name the name of the strategy
      * @param entryRule the entry rule
      * @param exitRule the exit rule
+     * @param unstablePeriod
      */
-    public BaseStrategy(Rule entryRule, Rule exitRule, int unstablePeriod) {
+    public BaseStrategy(String name, Rule entryRule, Rule exitRule, int unstablePeriod) {
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/BaseStrategy.java:89]: Please add old constructor `public BaseStrategy(Rule entryRule, Rule exitRule, int unstablePeriod)` that calls `this(null, entryRule,exitRule, unstablePerios)`

File: ta4j-core/src/main/java/org/ta4j/core/indicators/AwesomeOscillatorIndicator.java [hunk 4]
```diff
@@ -24,27 +24,50 @@ this software and associated documentation files (the "Software"), to deal in
...
+import org.ta4j.core.TimeSeries;
+import org.ta4j.core.indicators.helpers.MedianPriceIndicator;
 
 /**
  * Awesome oscillator. (AO)
- * <p></p>
- * @see <a href="http://www.forexgurus.co.uk/indicators/awesome-oscillator">http://www.forexgurus.co.uk/indicators/awesome-oscillator</a>
+ * <p>
+ * see https://www.tradingview.com/wiki/Awesome_Oscillator_(AO)
  */
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/AwesomeOscillatorIndicator.java]: Okay with replacing `<p></p>` but we should take `<p/>` than

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+		 * other-{@link Indicator indicator} decrease within the timeFrame. In
+		 * short: "other" and "ref" makes lower lows.
+		 */
+		negativeConvergent, 
+		
+		/**
+		 * Returns true for <b>"positiveDivergent"</b> when the values of the
+		 * ref-{@link Indicator indicator} increase and the values of the
+		 * other-{@link Indicator indicator} decrease within a timeFrame. In
+		 * short: "other" makes lower lows while "ref" makes higher lows.
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java]: `short: "other" makes lower lows while "ref" makes higher lows.`

Would it be better to say "...while "ref makes higher highs" ?

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	 * The <b>"minSlope"</b> is the minimum required slope for convergence or divergence
+	 * and must be a number between "0.1" and "1.0": <br/>
+	 * <br/>
+	 * 0.1: very unstrict<br/>
+	 * 0.3: strict (recommended) <br/>
+	 * 1.0: very strict <br/>
+	 * 
+	 * @param ref the indicator
+	 * @param other the other indicator
+	 * @param timeFrame
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java]: Please add at least simple description for parmeter e.g `@param timeFrame the time frame`

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+		 */
+		negativeDivergent,
+		
+		/**
+		 * Returns true for <b>"positiveConvergentStrict"</b> when the values of
+		 * the ref-{@link Indicator indicator} and the values of the
+		 * other-{@link Indicator indicator} increase consecutively within a
+		 * timeFrame. In short: "other" and "ref" makes strict higher highs.
+		 */
+		positiveConvergentStrict,
```
Comment (team172011) [inline ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java:89]: The user must be informed that if he uses one of the "**strict**" enum members, the minStrenght and minSlope will have no effect

File: ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java (hunk 6 above)
Comment (team172011) [reply ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/ConvergenceDivergenceIndicator.java]: Also for second constructor and the `@param index`

**Comment by nimo23** [comment]
Okay, I am finished! 

Thanks for your help!

**Comment by team172011** [comment]
thanks, too. I have added you to the AUTHORS file

**Pull request merged**